### PR TITLE
5 :arrow_right: 4 - EnumIface: suppress deprecation warning (#540)

### LINKED ISSUE #505: EnumIterator inherits from std::iterator base class which is deprecated in c++17
## Environment
* OS Version: macOS 12.6.5 (Monterey) with macOS 13.1 SDK
* Source or binary build? source build of gz-common5 as of 6f5ee941536e5c781bd826f6ce35360347d581ba (other versions too)

excerpt of `brew config`:

~~~
Clang: 14.0.0 build 1400
macOS: 12.6.5-x86_64
CLT: 14.2.0.0.1.1668646533
Xcode: N/A
~~~

## Description
* Expected behavior: package should compile without warnings
* Actual behavior: deprecation warning from `EnumIface.hh` when compiling `EnumIface_TEST.cc` since [`EnumIterator` inherits from `std::iterator`](https://github.com/gazebosim/gz-common/blob/gz-common5_5.3.1/include/gz/common/EnumIface.hh#L143-L145) and the `std::iterator` base class is [deprecated as of c++17](https://en.cppreference.com/w/cpp/iterator)

Screen Shot 2023-04-17 at 4 27 15 PM

~~~
src/EnumIface_TEST.cc:20:
/include/gz/common/EnumIface.hh:145:12: warning: 'iterator<std::bidirectional_iterator_tag, MyType>' is deprecated [-Wdeprecated-declarations]
    : std::iterator<std::bidirectional_iterator_tag, Enum>
           ^
src/EnumIface_TEST.cc:61:32: note: in instantiation of template class 'gz::common::EnumIterator<MyType>' requested here
  common::EnumIterator<MyType> end = MY_TYPE_END;
                               ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk/usr/include/c++/v1/__iterator/iterator.h:27:29: note: 'iterator<std::bidirectional_iterator_tag, MyType>' has been explicitly marked deprecated here
struct _LIBCPP_TEMPLATE_VIS _LIBCPP_DEPRECATED_IN_CXX17 iterator
                            ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk/usr/include/c++/v1/__config:1066:39: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#  define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                      ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk/usr/include/c++/v1/__config:1043:48: note: expanded from macro '_LIBCPP_DEPRECATED'
#    define _LIBCPP_DEPRECATED __attribute__ ((deprecated))
                                               ^
~~~

## Steps to reproduce

1. Build from source on macOS Monterey

or

1. Run https://build.osrfoundation.org/view/ign-garden/job/ignition_common-ci-gz-common5-homebrew-amd64 on one of the macOS-monterey nodes

## Output

[Build Status](https://build.osrfoundation.org/view/ign-garden/job/ignition_common-ci-gz-common5-homebrew-amd64/37/) https://build.osrfoundation.org/view/ign-garden/job/ignition_common-ci-gz-common5-homebrew-amd64/37/clang/

### PR DESCRIPTION
**Pull request by Crola1702**
# 🦟 Bug fix

Fixes #505 

## Summary
Backport of https://github.com/gazebosim/gz-common/pull/540

## Checklist
- [X] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: include/gz/common/EnumIface.hh
Comment (azeey) [inline include/gz/common/EnumIface.hh:145]: You'll need to include https://github.com/gazebosim/gz-utils/blob/ign-utils1/include/gz/utils/SuppressWarning.hh and use the `IGN` prefix. Unfortunately, these macros don't have a `GZ` prefixed alias in `ign-utils1`.

**Review by azeey — CHANGES_REQUESTED** [review]

**Comment by scpeters** [comment]
This change also required calling `target_link_library` on the utils library, which I've done in https://github.com/gazebosim/gz-common/commit/6652132af4f72217dfb30ac105d001a0a2be8b30 in a way that will minimize the diff when merging forward to `gz-common5`. Feel free to cherry-pick into this branch.

I think it is safe for our Ubuntu packages because the `libignition-common4-core-dev` already depends on `libignition-utils1-dev` (see [common4 control file](https://github.com/gazebo-release/gz-common4-release/blob/main/ubuntu/debian/control#L46))

File: src/CMakeLists.txt
Comment (azeey) [inline src/CMakeLists.txt:16]:
```suggestion
    ignition-utils${IGN_UTILS_VER }::ignition-utils${IGN_UTILS_VER }
```

**Review by azeey — CHANGES_REQUESTED** [review]

**Comment by azeey** [comment]
A version that doesn't use the ign-utils macros is in https://github.com/gazebosim/gz-common/pull/604. We can still merge this since the macros are better.

**Comment by Crola1702** [comment]
I think everything is ready to go. Waiting for approvals and then merge

**Comment by azeey** [comment]
The examples are failing on CI. This also shows that we are changing the dependency on ign-utils from being a requirement of the graphics component to being a requirement for the core library. I'm not exactly sure why the example is failing, but I'm inclined to close this PR and live with the bare pragmas that were added in #603.

**Pull request closed without merging**
